### PR TITLE
add torch.diff support for n greater than 1

### LINKED ISSUE #54853: `diff` higher order
Requesting feature.

```python
RuntimeError: diff only supports n = 1 currently. Please file an issue at 
https://github.com/pytorch/pytorch/issues/new?assignees=&labels=&template=feature-request.md 
if your use case requires supporting higher-order differences
```

Current workaround:

```python
def torch_diff(x, n, dim=-1):
    out = torch.diff(x, dim=dim)
    if n > 1:
        for _ in range(n - 1):
            out = torch.diff(out, dim=dim)
    return out
```

cc @mruberry @rgommers @heitorschueroff

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67260

Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -752,30 +752,41 @@ static inline void diff_check_compatible_shape(const Tensor& self, const c10::op
...
+    auto result = at::zeros_like(self);
+    result.copy_(self);
+    return result;
+  }
+
   auto out_len = self.size(dim) - 1;
-  if (self.dtype() == at::kBool) {
-    return at::logical_xor(at::narrow(self, dim, 1, out_len), at::narrow(self, dim, 0, out_len));
+  auto result = self;
+  bool is_kBool = self.dtype() == at::kBool;
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:771]: nit: `(self.dtype() == at::kBool;)` for readability

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -752,30 +752,41 @@ static inline void diff_check_compatible_shape(const Tensor& self, const c10::op
...
-      self.dim() >= 1,
+      self.dim()>= 1,
       "diff expects input to be at least one-dimensional");
 
   diff_check_compatible_shape(self, prepend, dim);
   diff_check_compatible_shape(self, append, dim);
 }
 
 static inline Tensor diff_helper(const Tensor& self, int64_t n, int64_t dim) {
+  if (n == 0){
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:763]: Comment here

Code formatting nit: `(n == 0) {`

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -752,30 +752,41 @@ static inline void diff_check_compatible_shape(const Tensor& self, const c10::op
...
-  if (self.dtype() == at::kBool) {
-    return at::logical_xor(at::narrow(self, dim, 1, out_len), at::narrow(self, dim, 0, out_len));
+  auto result = self;
+  bool is_kBool = self.dtype() == at::kBool;
+  n = n >= self.size(dim) ? self.size(dim) : n;
+
+  for (const auto i : c10::irange(n)) {
+    (void)i; //Suppress unused variable warning
+    if (is_kBool){
+      result = at::logical_xor(at::narrow(result, dim, 1, out_len), at::narrow(result, dim, 0, out_len));
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:777]: Nice bool handling

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -784,16 +795,34 @@ Tensor diff(const Tensor& self, int64_t n, int64_t dim, const c10::optional<Tens
 }
 
 static inline Tensor& diff_out_helper(const Tensor& self, int64_t n, int64_t dim, Tensor& result) {
-  auto out_len = self.size(dim) - 1;
+  if (n == 0){
+    at::native::resize_output(result, self.sizes());
+    check_scalar_type_device_layout_equal(result, self);
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:800]: This check is a little severe but that seems OK

File: aten/src/ATen/native/ReduceOps.cpp [hunk 5]
```diff
@@ -784,16 +795,34 @@ Tensor diff(const Tensor& self, int64_t n, int64_t dim, const c10::optional<Tens
...
+    result.copy_(self);
+    return result;
+  }
+
+  n = n >= self.size(dim) ? self.size(dim) : n;
+  auto out_len = self.size(dim) - n;
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:806]: nit: `const auto`

File: aten/src/ATen/native/ReduceOps.cpp [hunk 6]
```diff
@@ -784,16 +795,34 @@ Tensor diff(const Tensor& self, int64_t n, int64_t dim, const c10::optional<Tens
...
+  auto prev_result = self;
+
+  if (n > 1){
+    prev_result = diff_helper(self, n-1, dim);
+  }
+
   if (self.dtype() == at::kBool) {
-    return at::logical_xor_out(result, at::narrow(self, dim, 1, out_len), at::narrow(self, dim, 0, out_len));
+    at::logical_xor_out(result, at::narrow(prev_result, dim, 1, out_len), at::narrow(prev_result, dim, 0, out_len));
```
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:814]: Nice pattern of using `_out` variants for the final computation

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
         ((S, S), 0, (1, S), (2, S)),
         ((S, S), 0, None, (2, S)),
         ((S, S, S), 1, None, None),
-        ((S, S, S), 1, (S, 1, S), (S, 1, S)),)
+        ((S, S, S), 2, None, None),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2576]: Great extension

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
...
-                            low=None, high=None,
-                            requires_grad=requires_grad) if size_prepend else None,
-                make_tensor(size_append, device, dtype,
-                            low=None, high=None,
-                            requires_grad=requires_grad) if size_append else None)
-        sample_inputs.append(SampleInput(args[0], args=args[1:]))
+        prepend_size = 0 if (size_prepend is None) else size_prepend[dim]
+        append_size = 0 if (size_append is None) else size_append[dim]
+        dim_size = size[dim] + prepend_size + append_size
+        for n in range(dim_size + offset):  # test that there are no erors when n > dim_size
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2587]: This is very thorough (which is great!) but also computationally expensive (which is less great). How about including `n` in the sample input for each sample and just including 1-2 samples where `n > dim_size`?

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
...
+            args = (make_tensor(size, device, dtype,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2588]: nit: sometimes it's faster to partial the function like:

```
make_arg = partial(make_tensor, dict(device=device, dtype=dtype, requires_grad=requires_grad))
```

then these lines would be just 

```
make_arg(size_prepend) if size_prepend else None
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -2573,20 +2573,28 @@ def sample_inputs_diff(op_info, device, dtype, requires_grad, **kwargs):
...
+                                low=None, high=None,
+                                requires_grad=requires_grad), n, dim,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2590]: Hiding `n` and `dim` back here is tough to read. Passing them as kwargs would be more readable:

```
s = SampleInput(make_arg(size), kwargs={
  "n": n,
  "dim": dim,
  ...   
})
```

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -7920,6 +7928,11 @@ def ref_pairwise_distance(input1, input2):
                    safe_casts_outputs=True),
     OpInfo('diff',
            op=torch.diff,
+           # np.diff has np._NoValue as default values for prepend and append, compare_with_reference breaks if prepend/append
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7931]: Really nice reference handling

**Review by mruberry — APPROVED** [review]
Hey @mikaylagawarecki! This looks great. I left a few comments/suggestions for your review. Mostly to consider slimming the sample inputs so the tests don't take so long. 

I also added some additional tests, so let's wait for those to complete.

Thanks for patiently waiting for the review! Moving was a lot more time consuming than I expected. As soon as the tests pass this should be OK to land; ping me if you have any additional questions.

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31930294).

**Pull request closed without merging**